### PR TITLE
Partial Restriction of Site Access by IP

### PR DESCRIPTION
**Pull request by tloureiro**
I feel there are many ways to implement this, I'm not sure I've taken the best one. Happy to change the approach here.

### CODE REVIEW

File: 140-restricting-site-access/partial-restriction-by-ip.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Before trying to authenticate a user or displaying the login form, check if the request is coming from valid a valid IP
+ */
+add_action( 'wp_authenticate', 'vip__check_ips' );
+
+function vip__check_ips() {
+	if ( ! ( defined( 'WP_CLI' ) && WP_CLI ) ) {
+     // phpcs:ignore WordPressVIPMinimum.Variables.ServerVariables.UserControlledHeaders, WordPress.Security.ValidatedSanitizedInput.InputNotValidated, WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___SERVER__REMOTE_ADDR__
+		if ( ! in_array( $_SERVER['REMOTE_ADDR'], VIP__ALLOWED_IPS, true ) ) {
```
Comment (raamdev) [inline 140-restricting-site-access/partial-restriction-by-ip.php:27]: Can we check if `defined( VIP__ALLOWED_IPS ) && is_array( VIP__ALLOWED_IPS )` before using it?

File: 140-restricting-site-access/partial-restriction-by-ip.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+/**
+ * For every request, check if the request is made against a restricted path then check if the user is logged in
+ */
+function vip__check_restricted_pages() {
+
+	// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotValidated, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+	$current_path = $_SERVER['REQUEST_URI'];
+
+	$path_match = false;
+	foreach ( VIP_RESTRICTED_PATHS as $restricted_path ) {
```
Comment (raamdev) [inline 140-restricting-site-access/partial-restriction-by-ip.php:47]: Same thing here: Can we check if `defined( VIP_RESTRICTED_PATHS ) && is_array( VIP_RESTRICTED_PATHS )` before using it?

File: 140-restricting-site-access/partial-restriction-by-ip.php [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+define(
+	'VIP__RESTRICTED_PATHS',
+	array(
+		'/sample-page/',
+	)
+);
+
+
+/**
+ * Before trying to authenticate a user or displaying the login form, check if the request is coming from valid a valid IP
```
Comment (raamdev) [inline 140-restricting-site-access/partial-restriction-by-ip.php:20]: `from valid a valid IP` should be `from a valid IP`

**Review by raamdev — COMMENTED** [review]
@tloureiro 👋🏼  This looks great. I left a few inline comments regarding a few small things. In addition to the inline review comments, I have a couple of other questions:

- Has this been tested?
- Could we split this into two separate files for the two examples? e.g., `partial-restriction-by-ip.php` for the IP-based restriction example and `partial-restriction-by-uri.php` for the URI-based restriction example?

**Comment by tloureiro** [comment]
@raamdev thanks for reviewing it :dog: 

> @tloureiro 👋🏼 This looks great. I left a few inline comments regarding a few small things. In addition to the inline review comments, I have a couple of other questions:
> 
> * Has this been tested?

Yes, I've tested it locally.

> * Could we split this into two separate files for the two examples? e.g., `partial-restriction-by-ip.php` for the IP-based restriction example and `partial-restriction-by-uri.php` for the URI-based restriction example?

I'm not sure what would be the best way to do it. 

- We can keep certain IP's from logging in (this could be one example)
- We can keep non logged in users from accessing certain URL's (this could be another example)

But, from what I understand, the [page](https://docs.wpvip.com/technical-references/restricting-site-access/partial-restriction-site-access/) that needs code samples talks about these two things happening at the same time.

What do you think?

**Comment by raamdev** [comment]
> But, from what I understand, the page that needs code samples talks about these two things happening at the same time.

I think it would be good to have the examples in separate files so that we can more easily reference them. For example, the IP restriction example could be placed in a file called `partial-restriction-by-ip.php` and the path-based restriction example could be placed in a file called `partial-restriction-by-uri.php`. Then we can update our docs to describe the scenario where each technique can be used and then link to the examples individually.

**Review by raamdev — APPROVED** [review]
@tloureiro I did some testing and this looks good to me! As I noted above, I think it would be good to split the examples into two files for ease of reference, but the code itself looks good!

**Comment by tloureiro** [comment]
@raamdev I'm sorry I'm having a hard time understanding how the two examples would look like. Does this scenario make sense?
partial-restriction-by-ip.php (an example of blocking access to the whole site based on IP) (wouldn't this overlap with https://docs.wpvip.com/how-tos/block-requests/#h-block-by-ip#h-block-by-ip ?)
partial-restriction-by-uri.php (an example of blocking access to all users to a certain URL) (wouldn't this overlap with https://docs.wpvip.com/how-tos/block-requests/#h-block-by-other-logic ?)

**Comment by raamdev** [comment]
@tloureiro I was just thinking that there would be two files here in this repo:

#### 1 - *`partial-restriction-by-ip.php`*

Inside this file would be:

- `define('VIP__ALLOWED_IPS'...`
- `add_action( 'wp_authenticate', 'vip__check_ips' );`
- `function vip__check_ips() {...`

#### 2 - *`partial-restriction-by-uri.php`*

Inside this would be:

- `define( 'VIP__RESTRICTED_PATHS',...`
- `add_action( 'init', 'vip__check_restricted_pages' );`
- `function vip__check_restricted_pages() {`

This would allow us to reference the IP-based restriction example separately from the path-based restriction example. If the examples are in separate files, we can more easily link to each example.

**Comment by tloureiro** [comment]
@raamdev Thank you for the clarification. Done :+1:

**Review by raamdev — APPROVED** [review]
✅ Looks good to me! @tloureiro feel free to go ahead and merge this in when you're ready.

**Pull request merged**